packages/protocol/contracts/team/LibTokenGrant.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+library LibTokenGrant {
+    error INVALID_PARAMS();
+
+    function calcVestedAmount(
+        uint256 _grantAmount,
+        uint256 _vestDuration,
+        uint256 _time
+    )
+        internal
+        pure
+        returns (uint256)
+    {
+        if (_vestDuration == 0) return _grantAmount;
+        return _portion(_grantAmount, _time, _vestDuration);
+    }
+
+    function calcUnlockedAmount(
+        uint256 _grantAmount,
+        uint256 _vestDuration,
+        uint256 _unlockDuration,
+        uint256 _time
+    )
+        internal
+        pure
+        returns (uint256)
+    {
+        if (_vestDuration == 0 && _unlockDuration == 0) revert INVALID_PARAMS();
+
+        (uint256 a, uint256 b) = _vestDuration >= _unlockDuration
+            ? (_unlockDuration, _vestDuration)
+            : (_vestDuration, _unlockDuration);
+
+        if (a == 0) {
+            return _portion(_grantAmount, _time, b);
+        }
+
+        return _triagleArea(_grantAmount, a, b, _time, 0)
+            + _triagleArea(_grantAmount, a, b, _time, a + b)
+            - _triagleArea(_grantAmount, a, b, _time, a) //
+            - _triagleArea(_grantAmount, a, b, _time, b);
+    }
+
+    function _portion(uint256 z, uint256 t, uint256 tMax) private pure returns (uint256) {
+        if (t >= tMax) return z;
+        else return z * t / tMax;
+    }
+
+    function _triagleArea(
+        uint256 z,
+        uint256 a,
+        uint256 b,
+        uint256 t,
+        uint256 tMin
+    )
+        private
+        pure
+        returns (uint256)
+    {
+        if (a > b) revert INVALID_PARAMS();
+        if (t <= tMin) return 0;
+        uint256 _t = t - tMin;
+        return z * _t * _t / b / a / 2;
+    }
```

Comment:
Here is an example based on this math:

Bob grant is 100K tokens. It vests 4 years, and unlocks also 4 years.

Bob start date (grant start): 2023.01.01
At 2025.01.01 the math returns in `calcUnlockedAmount()` is the following:
```

       return _triagleArea(_grantAmount, a, b, _time, 0)
            + _triagleArea(_grantAmount, a, b, _time, a + b)
            - _triagleArea(_grantAmount, a, b, _time, a) //
            - _triagleArea(_grantAmount, a, b, _time, b);
```

Only first line returns a value, rest is 0, so let's just calculate the unlocked amount:

`return z * _t * _t / b / a / 2;`

100k * 2 * 2 / 4 / 4 /2 = 12.5K tokens.

So far so "good", seems kind of intentional.

So Bob started 2 years ago, an (now) at this point is able to withdraw 12.5% of his tokens.

Let see with different dates, how much is unlcoked at year 3 ?
100k * 3 * 3 / 4 / 4 / 2 = 28.125

 after 3 years spent this is 28.125%
 
 etc.